connect/runtime/src/test/java/org/apache/kafka/connect/integration/MonitorableSourceConnector.java

```diff
@@ -0,0 +1,152 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+
+/**
+ * A source connector to be used in integration tests. The rate of record production can be
+ * adjusted through the configs 'throughput' and 'messages.per.poll'
+ */
+public class MonitorableSourceConnector extends TestSourceConnector {
+    private static final Logger log = LoggerFactory.getLogger(MonitorableSourceConnector.class);
+
+    private String connectorName;
+    private ConnectorHandle connectorHandle;
+    private Map<String, String> commonConfigs;
+
+    @Override
+    public void start(Map<String, String> props) {
+        connectorHandle = RuntimeHandles.get().connectorHandle(props.get("name"));
+        connectorName = connectorHandle.name();
+        commonConfigs = props;
+        log.info("Started {} connector {}", this.getClass().getSimpleName(), connectorName);
+    }
+
+    @Override
+    public Class<? extends Task> taskClass() {
+        return MonitorableSourceTask.class;
+    }
+
+    @Override
+    public List<Map<String, String>> taskConfigs(int maxTasks) {
+        List<Map<String, String>> configs = new ArrayList<>();
+        for (int i = 0; i < maxTasks; i++) {
+            Map<String, String> config = new HashMap<>(commonConfigs);
+            config.put("connector.name", connectorName);
+            config.put("task.id", connectorName + "-" + i);
+            configs.add(config);
+        }
+        return configs;
+    }
+
+    @Override
+    public void stop() {
+        log.info("Stopped {} connector {}", this.getClass().getSimpleName(), connectorName);
+    }
+
+    @Override
+    public ConfigDef config() {
+        log.info("Configured {} connector {}", this.getClass().getSimpleName(), connectorName);
+        return new ConfigDef();
+    }
+
+    public static class MonitorableSourceTask extends SourceTask {
+        private String connectorName;
+        private String taskId;
+        private String topicName;
+        //TODO: extend its class and use it
+        private TaskHandle taskHandle;
+        private volatile boolean stopped;
+        private long startingSeqno;
+        private long seqno;
+        private long throughput;
+        private long batchSize;
+        private ThroughputThrottler throttler;
+
+        @Override
+        public String version() {
+            return "unknown";
+        }
+
+        @Override
+        public void start(Map<String, String> props) {
+            taskId = props.get("task.id");
+            connectorName = props.get("connector.name");
+            topicName = props.get("topic") != null ? props.get("topic") : "sequential-topic";
+            throughput = props.get("throughput") != null ? Long.valueOf(props.get("throughput")) : -1L;
+            batchSize = props.get("messages.per.poll") != null ? Long.valueOf(props.get("messages.per.poll")) : 1L;
```

Comment:
would `props.getOrDefault` work better in these cases?

Reply:
I believe I wanted to avoid that a key had been added with a `null` value, since we are not using `ConfigDef`s that avoid this situation. But I agree, that one should not add `null` even by accident, so I'll change to `getOrDefault`.